src/handler/TokenHandler.ts

```diff
@@ -80,18 +91,50 @@ class TokenHandler {
     ) {
       try {
         var resp = await this.getTokensFromRefreshToken();
+        idToken = resp.id_token;
       } catch (err) {
         // No refresh token in cookie or some other error
         return null;
       }
-      idToken = resp.id_token;
     }
     this.idToken = idToken;
     return new CotterIDToken(idToken);
   }
 
-  // Returns access token
+  // Refresh tokens only when needed.
   async getTokensFromRefreshToken(): Promise<OAuthToken> {
+    return new Promise<OAuthToken>((resolve, reject) => {
+      const checkTokenFetchProcess = () => {
+        if (this.tokenFetchingState === TOKEN_FETCHING_STATES.initial) {
+          this.getTokensFromRefreshTokenAPI();
+        } else if (
+          this.tokenFetchingState === TOKEN_FETCHING_STATES.ready &&
+          this.fetchTokenResp
+        ) {
+          resolve(this.fetchTokenResp);
+          return;
+        } else if (
+          this.tokenFetchingState === TOKEN_FETCHING_STATES.errorRetry
+        ) {
+          this.tokenFetchingState = TOKEN_FETCHING_STATES.initial;
+          reject();
```

Comment:
if it's errorRetry shouldn't it retry to fetch the token? why reject and return?

Reply:
errorRetry means if `getAccessToken` or `getIdToken` gets called again, can retry to refetch token. But we don't want to automatically refetch here cos it keeps on erroring with the same error then it'll keep on retrying forever

Reply:
I see, so the logic for retries should be handled by the user of the SDK?

Reply:
Yep, most probably would involve the end-user pressing smth